source/air/src/context.rs

```diff
@@ -314,6 +318,47 @@ impl Context {
         validity
     }
 
+    pub fn check_singular_valid(
+        &mut self,
+        vars: Vec<String>,
+        enss: Vec<Expr>,
+        reqs: Vec<Expr>,
+        s: &Span,
+    ) -> ValidityResult {
+        let nodes = singular_query_to_node(vars.to_vec(), enss.to_vec(), reqs.to_vec());
+        self.air_initial_log.log_node(&nodes);
+        self.air_middle_log.log_node(&nodes);
+        self.air_final_log.log_node(&nodes);
+
+        let query = singular_printer(vars, enss, reqs);
+        self.air_initial_log.comment(&query);
+        self.air_middle_log.comment(&query);
+        self.air_final_log.comment(&query);
```

Comment:
I'm confused as what's happening here? Is this appending the singular query as a comment to the air file?
If so, why have the vars, requires, and ensures been printed as s-expr nodes? Is z3 supposed to see them?

Reply:
I thought air log as an incorporated log file for all solvers, and imagined invoking Singular when `singular-check-valid` is encountered.  Would it be better to log the Singular part into a separate file?